### PR TITLE
[Serve] Support basic Starlette response types

### LINKED ISSUE #11575: [Serve] Full Support for HTTP response
_Originally posted by @TanjaBayer in https://github.com/ray-project/ray/discussions/11570_

Allow configurable response type:
- HTTP status code
- Content Type
- Header
- Multipart Encoding

cc @edoakes @architkulkarni

### LINKED ISSUE #12299: [serve] Support Starlette Response type

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?

This PR allows basic [Starlette response types](https://www.starlette.io/responses/) to be returned by Ray Serve backends.  This allows for more customizable HTTP responses, including response codes, cookies, content-type, and so on.  A test is included for the examples for `Response`, `HTMLResponse`, `JSONResponse`, `PlainTextResponse` and `RedirectResponse` from the documentation linked above.

For now, non-serializable Starlette Response types such as `StreamingResponse` are unsupported, and an error is raised if a `StreamingResponse` is returned by a backend function or class.

This PR also adds starlette as a dependency for Serve and for the docs.

## Related issue number

Closes https://github.com/ray-project/ray/issues/11575.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -222,7 +222,8 @@ def create_backend(
         Args:
             backend_tag (str): a unique tag assign to identify this backend.
             func_or_class (callable, class): a function or a class implementing
-                __call__.
+                __call__, returning a JSON-serializable object or a
+                serializable Starlette Response object.
```
Comment (edoakes) [inline python/ray/serve/api.py:226]: All Starlette response objects should be serializable, right? If so, can drop the qualifier - it'll probably confuse people

File: python/ray/serve/http_proxy.py [hunk 2]
```diff
@@ -126,6 +127,16 @@ async def __call__(self, scope, receive, send):
...
             await error_sender(error_message, 500)
+        elif isinstance(result, starlette.responses.Response):
+            if isinstance(result, starlette.responses.StreamingResponse):
+                raise TypeError("Starlette StreamingResponse returned by "
+                                f"backend for endpoint {endpoint_name}. "
+                                "StreamingResponse is unserializable and not "
+                                "supported by Ray Serve.  Consider using "
+                                "another Starlette response type such as "
+                                "Response, HTMLResponse, PlainTextResponse, "
+                                "or JSONResponse.")
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:138]: Can you add a line that requests them to file an issue on github if this is a desired feature?

File: python/ray/serve/http_proxy.py [hunk 3]
```diff
@@ -126,6 +127,16 @@ async def __call__(self, scope, receive, send):
...
+            await result(scope, receive, send)
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:139]: oh, wow, so this was literally one line of code?!

**Review by edoakes — APPROVED** [review]
LGTM, just a few small comments. Please make sure to close relevant issue and notify users who asked about it

**Comment by edoakes** [comment]
Lint failing FYI: https://travis-ci.com/github/ray-project/ray/jobs/458778170

**Comment by edoakes** [comment]
@architkulkarni there are some test failures:
https://travis-ci.com/github/ray-project/ray/jobs/458972916

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
cc @TanjaBayer, exciting work happening!

File: python/ray/serve/http_proxy.py (hunk 3 above)
Comment (simon-mo) [reply python/ray/serve/http_proxy.py:141]: yeah :-)

**Pull request merged**